### PR TITLE
[11.x] Allows to persist database connection between tests and only reset PDO in `afterClassTearDown()`

### PR DESCRIPTION
**Pull request by crynobone**
#### Background

PDO connection doesn't automatically get closed when the application is unset but instead requires explicitly set to `null` to trigger connection close:

CleanShot 2023-12-15 at 15 35 46

https://www.php.net/manual/en/pdo.connections.php

Since PHPUnit is especially a long-running process when the application gets booted and terminated on each test the PDO may persist and eventually can cause max database connections. To solve we have two options:

#### Option A: We can explicitly disconnect the database between tests

Disadvantages:
* Caused breaking change to the existing application since it is still possible to use PDO and loaded instances (without IoC) after `Iluminate\Foundation\Testing\TestCase::tearDown()` has been called.
* Any cached/static instances holding `PDO` will still prevent the database from being disconnected.

#### Option B: Reuse `PDO` instances between tests (provided in this PR)

Disadvantage:

* The only issue I can think of if tests within a TestCase change database configuration between tests, since the PDO instance is loaded using cache it not going to fetch the specific configuration defined for the test. However, I don't see any reason why a developer would have this for the application tests, even for package development this would be a rare requirement.

To cover Option A, you can configure a base `TestCase` and set `$disconnectDatabaseConnections` property to `true`:

```diff
<?php

namespace Tests;

abstract class TestCase extends \Illuminate\Foundation\Testing\TestCase
{
    use CreatesApplication;

+   protected $disconnectDatabaseConnections = true;

}
```

### CODE REVIEW

File: src/Illuminate/Database/Connectors/ConnectionFactory.php [hunk 1]
```diff
@@ -39,7 +40,7 @@ public function __construct(Container $container)
      * @param  string|null  $name
      * @return \Illuminate\Database\Connection
      */
-    public function make(array $config, $name = null)
+    public function make(array $config, $name)
```
Comment (crynobone) [inline src/Illuminate/Database/Connectors/ConnectionFactory.php:43]: `$name` shouldn't accept `null` since `parseConfig()` requires `$name` only to be `string`.

**Comment by mpyw** [comment]
Related to #49373? Please comment there

File: src/Illuminate/Foundation/Testing/DatabaseConnectionFactory.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            return static::$cachedConnections[$key] = $this->factory->make($config, $name);
+        }
+
+        $config = $this->parseConfig($config, $name);
+
+        $connection = $this->createConnection(
+            $config['driver'], static::$cachedConnections[$key]->getRawPdo(), $config['database'], $config['prefix'], $config
+        )->setReadPdo(static::$cachedConnections[$key]->getRawReadPdo());
+
+        return static::$cachedConnections[$key] = $connection;
```
Comment (SakiTakamachi) [inline src/Illuminate/Foundation/Testing/DatabaseConnectionFactory.php:61]:
```suggestion
        static::$cachedConnections[$key]->disconnect();
        return static::$cachedConnections[$key] = $this->factory->make($config, $name);
```

[IMHO]
I'm concerned that reusing connections may have unintended side effects. How about trying disconnecting and reconnecting?

Comment (crynobone) [reply src/Illuminate/Foundation/Testing/DatabaseConnectionFactory.php:61]: The whole point of this PR is to reuse available PDO instances, this is inline with what's already approved to in-memory database feature available in `RefreshDatabase` for Laravel 11.

Comment (SakiTakamachi) [reply src/Illuminate/Foundation/Testing/DatabaseConnectionFactory.php:61]: There is a risk that the transaction state and session settings will be carried over as well, making it difficult to understand why the test failed.What do you think about this?

These are just the cases I can think of right now, so I may have missed something.

Comment (crynobone) [reply src/Illuminate/Foundation/Testing/DatabaseConnectionFactory.php:61]: Wouldn't that already been handled by `DatabaseTransactionManager`? Can you provide a use case where `DB::transaction()` is run without being reset to 0?

**Comment by KentarouTakeda** [comment]
Since this applies to code that falls outside the scope of *Suggested change* , I would like to comment on it separately. (I'm not familiar with MySQL, so I'd like to comment only on PostgreSQL. But I think MySQL probably has similar features as well.)

https://github.com/laravel/framework/pull/49385#discussion_r1427684362

> There is a risk that the transaction state and session settings will be carried over

There is some changeable state that persists while a session is open, **whether inside or outside a transaction, and whether commit or rollback**.

However, certain states also allow us to explicitly discard their state.
(Some implementations may already exist. Sorry if there are.)

* [Session Authorization](https://www.postgresql.org/docs/16/sql-set-session-authorization.html)
  * Return to initial state with `RESET SESSION AUTHORIZATION`.
* [Connection status settings](https://www.postgresql.org/docs/16/sql-set.html).
  * `RESET ALL` resets everything. (Also valid for many settings other than `SET` statements)
  * PHP's `ext/pgsql` extension [uses this method](https://github.com/php/php-src/blob/php-8.3.0/ext/pgsql/pgsql.c#L618) to reset the state of persistent connections. I think it's a good idea to implement the same in our framework.
* [Advisory lock](https://www.postgresql.org/docs/16/functions-admin.html#FUNCTIONS-ADVISORY-LOCKS), [Named Prepared Statement](https://www.postgresql.org/docs/16/sql-prepare.html)
  * All will be discarded with `DISCARD ALL` .

In order to preserve existing behavior, we should execute these (or other?) statements to return the session state to its initial state as much as possible.

File: src/Illuminate/Foundation/Testing/DatabaseConnectionFactory.php (hunk 2 above)
Comment (SakiTakamachi) [reply src/Illuminate/Foundation/Testing/DatabaseConnectionFactory.php:61]: If using two-phase commit, it can easily happen because Laravel does not control the transaction.

Also, the problem is more with the session settings, and when the transaction isolation level is changed, the level is reflected in the next test as well.

Comment (crynobone) [reply src/Illuminate/Foundation/Testing/DatabaseConnectionFactory.php:61]: So all this issue should also exist in Laravel Octane, right?

Comment (SakiTakamachi) [reply src/Illuminate/Foundation/Testing/DatabaseConnectionFactory.php:61]: Probably not. The PDO instance you get should be different each time since the DI container is reset each time.

(I haven't really looked into how 8.x handles database connections yet, so I could be wrong.)

File: tests/Integration/Database/SchemaBuilderTest.php [hunk 3]
```diff
@@ -26,6 +26,8 @@ public function testDropAllTables()
 
         Schema::dropAllTables();
 
+        DB::reconnect();
```
Comment (crynobone) [inline tests/Integration/Database/SchemaBuilderTest.php:29]: This is only required to fixed `sqlite` and not using in-memory database. It seem using `Schema::dropAllTables()` require a new connection on GA, unable to reproduce this locally

File: src/Illuminate/Foundation/Testing/DatabaseConnectionFactory.php [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+     * @param  array  $config
+     * @param  string  $name
+     * @return \Illuminate\Database\Connection
+     */
+    #[\Override]
+    public function make(array $config, $name)
+    {
+        $key = $name ?? $config['name'];
+
+        // SQLite doesn't have any max connections limitation so it should be safe to just create a new connection between tests.
```
Comment (KentarouTakeda) [inline src/Illuminate/Foundation/Testing/DatabaseConnectionFactory.php:46]: *IMO:*
```suggestion
        // In-Memory Databases doesn't have any max connections limitation so it should be safe to just create a new connection between tests.
        // Because some tests may be depend on thier volatile, we should always create new connections to avoid carrying over previous data.
```

In addition to the current comment,
How about adding that it supports testing assuming the volatility of the In Memory Database?

I initially tried turning this off, thinking it would be better to have consistent behavior regardless of the driver, but many of the tests for the conditions mentioned above failed. So, I understand that this process is necessary from a different perspective than limiting the number of connections.

Comment (crynobone) [reply src/Illuminate/Foundation/Testing/DatabaseConnectionFactory.php:46]: The affect of this behaviour is a huge breaking change for `:memory:` and should only occur with `RefreshDatabase` new improvements for L11.

Comment (crynobone) [reply src/Illuminate/Foundation/Testing/DatabaseConnectionFactory.php:46]: See https://github.com/laravel/framework/pull/47912

Comment (KentarouTakeda) [reply src/Illuminate/Foundation/Testing/DatabaseConnectionFactory.php:46]: I caught up, thank you.

In addition to the general case you mention, there were some parts of Laravel's own testing that were related to this. for example:

https://github.com/laravel/framework/blob/41c572ef01841be07304b87c3926a652bddc1d59/tests/Integration/Database/DatabaseSchemaBlueprintTest.php#L35-L38
(This test fails with `table "users" already exists` if we persist the connection.)

Either way, the meaning of the comment is essentially the same, so either is fine. Thank you for confirmation!

**Comment by taylorotwell** [comment]
> Since PHPUnit is especially a long-running process when the application gets booted and terminated on each test the PDO may persist and eventually can cause max database connections. 

What causes it to persist? How do you recreate this on a fresh Laravel application to hit the max connections error?

**Comment by KentarouTakeda** [comment]
> What causes it to persist?

Mentioned in https://github.com/laravel/framework/discussions/49373#discussioncomment-7854320. It appears that a reference to the connection resource is being held in some code somewhere, but I don't know where. It may be Laravel code, but it may also be processed by PHPUnit.

> How do you recreate this on a fresh Laravel application to hit the max connections error?

Mentioned in #49311. It happens by simply repeating the test case.

**Comment by crynobone** [comment]
> What causes it to persist? How do you recreate this on a fresh Laravel application to hit the max connections error?

### Before 

CleanShot 2023-12-17 at 16 59 48

We assume using `flush()` and `unset($app)` would be enough to clear PDO reference but it doesn't see so (which let me to believe that the documentation in PHP meant that we do need to explicitly set PDO instances to `null`.

### After

CleanShot 2023-12-17 at 16 59 57

**Comment by crynobone** [comment]
We tried disconnecting the database connection in tearDown via #49327 but this caused a regression (reported via #49359) and proven that you can still interacts with PDO after `unset($app)`

**Comment by taylorotwell** [comment]
Sorry, but if the current behavior is a significant problem, why are we not seeing **many more** reports of this from basically every Laravel application in the world that has tests?

**Comment by mpyw** [comment]
@taylorotwell Possibly because everyone uses `RefreshDatabase`. But in some reason I prefer `DatabaseTruncation`, so I critically encountered the problem.

**Comment by SakiTakamachi** [comment]
@taylorotwell 

Here are simple steps to reproduce this problem.

The following test code:
```
<?php

namespace Tests\Feature;

use Tests\TestCase;
use Illuminate\Foundation\Testing\DatabaseMigrations;
use Illuminate\Foundation\Testing\DatabaseTruncation;

class ConnTest extends TestCase
{
    use DatabaseMigrations;
    // or use DatabaseTruncation;
    public static function repeat(): array
    {
        return array_fill(0, 10, []);
    }

    /** @dataProvider repeat */
    public function test(): void
    {
        echo "waiting input...";
        ob_flush();
        fgets(STDIN);
        echo "input received\n";
        $this->assertTrue(true);
    }
}

```

run test:
```
php artisan test tests/Feature/ConnTest.php
```

The test will stop until the Enter key is pressed:
```
waiting input...
input received
waiting input...
input received
waiting input...

...(omit)
```

For MySQL, you can check the number of connections with the following query:
```
show status like 'Threads_connected';
```

If you check the number of connections while stopping the test, you will see that the number of connections increases as the test case progresses:
```
mysql> show status like 'Threads_connected';
+-------------------+-------+
| Variable_name     | Value |
+-------------------+-------+
| Threads_connected | 2     |
+-------------------+-------+
1 row in set (0.00 sec)

mysql> show status like 'Threads_connected';
+-------------------+-------+
| Variable_name     | Value |
+-------------------+-------+
| Threads_connected | 3     |
+-------------------+-------+
1 row in set (0.00 sec)

...(omit)
```

**Comment by KentarouTakeda** [comment]
I found the reason for @taylorotwell's question.

In short, with the current implementation, whether the problem occurs or not depends on when PHP's circular reference collector is triggered.

It varies depending on the PHP settings and the memory capacity of the execution environment, so

* [The code](https://github.com/laravel/framework/issues/49359) shown by @rikvdh may not always work.
* The problem that @mpyw actually encountered may not be reproduced depending on the environment.

In any case, the current implementation's behavior changes depending on the environment (or luck), and this pull request will resolve that.

I have prepared CI results in my repository [that can confirm the above operation](https://github.com/KentarouTakeda/example-number-of-db-connections/blob/main/tests/Feature/RecordTheNumberOfDBConnectionsTest.php). Complete results can be found [here](https://github.com/KentarouTakeda/example-number-of-db-connections/actions/runs/7243754950/job/19731056620#step:6:23).

connection-graph

This image visualizes the increase and decrease in the number of simultaneous connections by repeating the same test. `$all->flush()` is called every time, but the actual timing of disconnection is different each time.

Complete CI results also include cases where we explicitly call `gc_collect_cycles()`. Since the number of connections in this case is constant, we can see that the trigger for disconnection is GC.

File: composer.json [hunk 5]
```diff
@@ -106,7 +106,7 @@
         "league/flysystem-sftp-v3": "^3.0",
         "mockery/mockery": "^1.5.1",
         "nyholm/psr7": "^1.2",
-        "orchestra/testbench-core": "^9.0",
+        "orchestra/testbench-core": "dev-persist-db-connections",
```
Comment (driesvints) [inline composer.json:109]: @crynobone should we put this PR in draft while this dependency isn't ready yet?

Comment (crynobone) [reply composer.json:109]: I can't merge Testbench Core branch to stable unless Laravel commit with this changes it Testbench will depends on the new `Illuminate\Foundation\Testing\DatabaseConnectionFactory` added in this PR.

File: tests/Integration/Foundation/Console/AboutCommandTest.php [hunk 6]
```diff
@@ -3,10 +3,12 @@
 namespace Illuminate\Tests\Integration\Foundation\Console;
 
 use Illuminate\Testing\Assert;
+use Orchestra\Testbench\Attributes\WithEnv;
 use Orchestra\Testbench\TestCase;
 
 use function Orchestra\Testbench\remote;
 
+#[WithEnv('APP_MAINTENANCE_STORE', 'array')]
```
Comment (crynobone) [inline tests/Integration/Foundation/Console/AboutCommandTest.php:11]:
```suggestion
```

File: tests/Integration/Foundation/Console/AboutCommandTest.php [hunk 7]
```diff
@@ -3,10 +3,12 @@
 namespace Illuminate\Tests\Integration\Foundation\Console;
 
 use Illuminate\Testing\Assert;
+use Orchestra\Testbench\Attributes\WithEnv;
```
Comment (crynobone) [inline tests/Integration/Foundation/Console/AboutCommandTest.php:6]:
```suggestion
```

File: tests/Integration/Foundation/MaintenanceModeTest.php [hunk 8]
```diff
@@ -11,10 +11,12 @@
 use Illuminate\Support\Carbon;
 use Illuminate\Support\Facades\Event;
 use Illuminate\Support\Facades\Route;
+use Orchestra\Testbench\Attributes\WithEnv;
```
Comment (crynobone) [inline tests/Integration/Foundation/MaintenanceModeTest.php:14]:
```suggestion
```

File: tests/Integration/Foundation/MaintenanceModeTest.php [hunk 9]
```diff
@@ -11,10 +11,12 @@
...
 use Orchestra\Testbench\Http\Middleware\PreventRequestsDuringMaintenance as TestbenchPreventRequestsDuringMaintenance;
 use Orchestra\Testbench\TestCase;
 use Symfony\Component\HttpFoundation\Cookie;
 
+#[WithEnv('APP_MAINTENANCE_DRIVER', 'file')]
```
Comment (crynobone) [inline tests/Integration/Foundation/MaintenanceModeTest.php:19]:
```suggestion
```

**Comment by deleugpn** [comment]
I'd like to chime in on this since it **seems like** it may heavily affect me.

> The only issue I can think of if tests within a TestCase change database configuration between tests, since the PDO instance is loaded using cache it not going to fetch the specific configuration defined for the test. However, I don't see any reason why a developer would have this for the application tests, even for package development this would be a rare requirement.

The reason I have is because I run a multi-database multi-tenant application and sometimes I have some critical features that need the extra level of security to ensure that:

1) Given 2 tenants (2 databases) and 1 user in each tenant 
2) When User A performs an action
3) It reflects on Tenant's database without crossing/impacting the tenant boundary.

------

From reading the PR, I see Taylor's point. I've been using Laravel since 5.0 and I have never had any max connection issues. I believe **many** Laravel users don't face this issue at all. Making a significantly drastic change to how the framework handles database connections affecting **every Laravel/Phpunit Users** to fix an edge case without a way to keep the existing behavior seems like an opportunity to invite many more issues/bug reports than the current situation.

I'm all for reusing the same PDO connection across many tests which could improve performance and speed up tests, but I feel like users need to be able to better control this. 

While we're on this subject, a paper cut issue I've had with Laravel for many years is precisely how DatabaseManager is tied with Config Repository, making it unable to work without global state changes that causes confusion.

image

image

File: src/Illuminate/Database/Connectors/ConnectionFactory.php (hunk 1 above)
Comment (deleugpn) [reply src/Illuminate/Database/Connectors/ConnectionFactory.php:43]: `parseConfig()` is only restricted via doc-block documentation. Since the property isn't really strictly typed, it actually works fine by using `$factory->make($config)` on userland code.

**Comment by deleugpn** [comment]
> @taylorotwell Possibly because everyone uses `RefreshDatabase`. But in some reason I prefer `DatabaseTruncation`, so I critically encountered the problem.

@mpyw If this problem mainly affects `DatabaseTruncation`, can't it be fixed inside `DatabaseTruncation` trait?

**Comment by crynobone** [comment]
> I believe many Laravel users don't face this issue at all. 

It all depends on the max connection configured locally and on CI, we did face this issue on Framework own tests a few years back: https://github.com/laravel/framework/pull/39877

> I'm all for reusing the same PDO connection across many tests which could improve performance and speed up tests, but I feel like users need to be able to better control this.

Welcome any suggestion on this, we did add ability to disconnect the database connection via a property flag in this PR.

>  If this problem mainly affects DatabaseTruncation, can't it be fixed inside DatabaseTruncation trait?

If the approach is to disconnect all database during tearDown, this have been proposed and rejected, also potentially impacts performance (since we only disconnect database connectio via `RefreshDatabase` and `DatabaseMigrations` trait at the moment) with less control.

File: src/Illuminate/Database/Connectors/ConnectionFactory.php (hunk 1 above)
Comment (crynobone) [reply src/Illuminate/Database/Connectors/ConnectionFactory.php:43]: For multi-tenant applications where you may want to change the default connections, this might not be the best usage:

CleanShot 2024-01-02 at 10 24 39

**Comment by taylorotwell** [comment]
Would it be simpler and less breaking (particularly for the multi-tenant scenario) to just call `gc_collect_cycles` periodically while running tests?

File: src/Illuminate/Database/Connectors/ConnectionFactory.php (hunk 1 above)
Comment (deleugpn) [reply src/Illuminate/Database/Connectors/ConnectionFactory.php:43]: sorry I didn't see your reply here, unfortunately I don't get notifications on the Laravel repository anymore (GitHub bug).

Whether you change or not the default connection in that script it's irrelevant since the configuration is being explicitly passed to the `$factory->make()`. The variable you hold is the one that is relevant to which database connection you have.

the `getName()` method is returning exactly what was provided when constructing the connection (`null`), which is my point about this breaking change. It works as intended and it doesn't seem relevant to break it

**Comment by deleugpn** [comment]
> If the approach is to disconnect all database during tearDown, this have been proposed and rejected, also potentially impacts performance (since we only disconnect database connectio via `RefreshDatabase` and `DatabaseMigrations` trait at the moment) with less control.

I guess my primary point is: I'm a user of the `DatabaseTransactions` trait and as such it seems I'm more unlikely to face this problem than users of the `DatabaseTruncation` as it was mentioned. 

Perhaps the implemented solution to the problem can be more targeted at where the problem is and leave users that are outside the scope of it unaffected. The problem with the current proposal is that it completely hijacks the DatabaseConnectionFactory at a global level (`OverrideProvidersForTesting`), causing 100% of Laravel installation to be affected by this change regardless of whether they are subject to the reported bug or not.

**Comment by mpyw** [comment]
@crynobone 

> Would it be simpler and less breaking (particularly for the multi-tenant scenario) to just call gc_collect_cycles periodically while running tests?

Is this the answer?
